aiaccel/hpo/samplers/nelder_mead_sampler.py

```diff
@@ -0,0 +1,129 @@
+from __future__ import annotations
+
+import warnings
+from collections.abc import Sequence
+from typing import Any
+
+import numpy as np
+import optuna
+from optuna.distributions import BaseDistribution
+from optuna.study import Study
+from optuna.trial import FrozenTrial, TrialState
+
+from aiaccel.hpo.algorithms import NelderMeadAlgorism, NelderMeadCoefficient, NelderMeadEmptyError
+
+__all__ = ["NelderMeadSampler", "NelderMeadEmptyError"]
+
+
+class NelderMeadSampler(optuna.samplers.BaseSampler):
+    def __init__(
+        self,
+        search_space: dict[str, tuple[float, float]],
+        seed: int | None = None,
+        rng: np.random.RandomState | None = None,
+        coeff: NelderMeadCoefficient | None = None,
+        block: bool = False,
+        sub_sampler: optuna.samplers.BaseSampler | None = None,
+    ) -> None:
+        self._search_space = search_space
+        _rng = rng if rng is not None else np.random.RandomState(seed) if seed is not None else None
+
+        self.nm = NelderMeadAlgorism(
+            search_space=self._search_space,
+            coeff=coeff,
+            rng=_rng,
+            block=block,
+        )
+        self.sub_sampler = sub_sampler
+
+    def infer_relative_search_space(self, study: Study, trial: FrozenTrial) -> dict[str, BaseDistribution]:
+        return {}
+
+    def sample_relative(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        search_space: dict[str, BaseDistribution],
+    ) -> dict[str, Any]:
+        return {}
+
+    def before_trial(self, study: Study, trial: FrozenTrial) -> None:
+        if "fixed_params" in trial.system_attrs:  # enqueued trial
+            fixed_params = trial.system_attrs["fixed_params"]
+            if fixed_params.keys() != self._search_space.keys():
+                raise RuntimeError("All parameters must be given when executing enqueue_trial.")
+
+            params = np.array([fixed_params[name] for name in self._search_space])
+        else:
+            while True:
+                try:
+                    params = self.nm.get_vertex()
+                except NelderMeadEmptyError as e:
+                    if self.sub_sampler is None:
+                        raise e
+                    else:
+                        self.sub_sampler.before_trial(study, trial)
+                        trial.set_user_attr("sub_trial", True)
+                        break
+
+                if all(low < x < high for x, (low, high) in zip(params, self._search_space.values(), strict=False)):
+                    break
+
+                self.nm.put_value(params, np.inf)
```

Comment:
```suggestion
            try:
                while True:
                    params = self.nm.get_vertex()

                    if all(low < x < high for x, (low, high) in zip(params, self._search_space.values(), strict=False)):
                        break

                    self.nm.put_value(params, np.inf)
            except NelderMeadEmptyError as e:
                if self.sub_sampler is None:
                    raise e
                else:
                    self.sub_sampler.before_trial(study, trial)
                    trial.set_user_attr("sub_trial", True)
```
こう書くとロジックがシンプルに見える気がしました．